### ISSUE TITLE
LDAP auth: Cannot establish TLS1.2 connection with Active Directory server using SHA-512 certificates

### ISSUE THREAD

**Issue by glennmcallister** [hashicorp/vault#2058]
**Vault Release**: 0.6.2

**Reproduction Steps**:

1. Setup Active Directory DC with TLS, using a SHA-512 hash cert.  Yep, this is a pain.
2. Setup the LDAP auth method to communicate with this DC.
3. Attempt to authenticate.

**Expected behaviour**: Successful authentication.

**Actual behaviour**: Connection failure.

```
$ vault auth --method=ldap username=glenn.mcallister
Password (will be hidden):
Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/ldap/login/glenn.mcallister
Code: 400. Errors:

* cannot connect to LDAP: LDAP Result Code 200 "": read tcp #.#.#.#:41496->#.#.#.#:636: read: connection reset by peer
```

NOTE: In the above scenario, I'm proxying the connection from localhost to my server.

To expand upon step 2 above, my configuration (with sensitive data elided) looks like:

```
$ vault read auth/ldap/config
Key            	Value
---            	-----
binddn         	<snip>
bindpass       	<snip>
certificate    	<snip>
discoverdn     	false
groupattr      	memberof
groupdn        	OU=Cloud Ops,DC=bluecatlabs,DC=net
groupfilter    	(&(objectClass=group)(member:1.2.840.113556.1.4.1941:={{.UserDN}}))
insecure_tls   	true
starttls       	false
tls_min_version	tls12
upndomain      	bluecatlabs.net
url            	ldaps://<snip>.bluecatlabs.net
userattr       	samaccountname
userdn         	DC=bluecatlabs,DC=net

The following warnings were returned from the Vault server:
* Read access to this endpoint should be controlled via ACLs as it will return the configuration information as-is, including any passwords.
```

We narrowed the actual problem down to what's described [here](https://social.technet.microsoft.com/Forums/en-US/b6ffa278-4a04-4609-ac35-8390f5ba9cb6/ldap-over-ssl-on-windows-2012r2-server-dcs-tls-12-not-working?forum=winserversecurity). The relevant part is the following:

> Looking at the list of cipher suites sent by LDP.exe, none included SHA512 as a signature algorithm.  Moreover, it doesn't appear that any implementation of TLS supports SHA512 either.  This is an issue if your certificates are signed with SHA512, since per RFC 5246 (7.4.2), "If the client provided a 'signature_algorithms' extension, then all certificates provided by the server MUST be signed by a hash/signature algorithm pair that appears in that extension."

When you look at the ciphers provided by golang's `cypher/tls` implementation, none of them are 512. 

Our administrator is not going create a new certificate chain at SHA-256. Full stop.

When I created a client using `go-ldap`, I **was** able to connect, when I set `tsl.Config{MaxVersion: tls.VersionTLS11}`. Unfortunately, we can't set this value via the Vault LDAP configuration currently.

**Comment by jefferai** [comment]
You could try bumping the TLS min version in the LDAP backend down from tls12 to tls11. You may also want to set starttls to true.

**Comment by glennmcallister** [comment]
`starttls` is required only if you first try to connect to an `ldap://` endpoint and are upgrading to `ldaps://`. 

With `tls_min_version=tls11`:

```
$ vault auth --method=ldap username=glenn.mcallister
Password (will be hidden):
Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/ldap/login/glenn.mcallister
Code: 400. Errors:

* cannot connect to LDAP: LDAP Result Code 200 "": read tcp #.#.#.#:46078->#.#.#.#:636: read: connection reset by peer
```

With both values set:

```
$ vault auth --method=ldap username=glenn.mcallister
Password (will be hidden):
Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/ldap/login/glenn.mcallister
Code: 400. Errors:

* cannot connect to LDAP: LDAP Result Code 200 "": read tcp #.#.#.#:46158->#.#.#.#:636: read: connection reset by peer
```

My config:

```
mp15gmcalliste1:~ gmcallister$ vault read auth/ldap/config
Key            	Value
---            	-----
binddn         	<snip>
bindpass       	<snip>
certificate    	<snip>
discoverdn     	false
groupattr      	memberof
groupdn        	OU=Cloud Ops,DC=bluecatlabs,DC=net
groupfilter    	(&(objectClass=group)(member:1.2.840.113556.1.4.1941:={{.UserDN}}))
insecure_tls   	true
starttls       	true
tls_min_version	tls11
upndomain      	bluecatlabs.net
url            	ldaps://<snip>.bluecatlabs.net
userattr       	samaccountname
userdn         	DC=bluecatlabs,DC=net

The following warnings were returned from the Vault server:
* Read access to this endpoint should be controlled via ACLs as it will return the configuration information as-is, including any passwords.
```

On the server side, we see reams of the following schannel events:

**Event 36888**:
```
A fatal alert was generated and sent to the remote endpoint. This may result in termination of the connection. The TLS protocol defined fatal error code is 40. The Windows SChannel error state is 1205.
```

**Event 36874**:
```
An TLS 1.1 connection request was received from a remote client application, but none of the cipher suites supported by the client application are supported by the server. The SSL connection request has failed.
```

**Comment by glennmcallister** [comment]
I apologize, I realized I missed a rather important point in our configuration. The AD server is Server 2012 R2 Datacenter.

**Comment by jefferai** [comment]
What version of Go did you compile your testing go-ldap program against?

**Comment by glennmcallister** [comment]
Same version in both OS X and Linux:

**OS X:**
```
$ which go
/usr/local/go/bin/go
$ go version
go version go1.7 darwin/amd64
```

**CentOS:**
```
$ which go
/usr/local/go/bin/go
$ go version
go version go1.7 linux/amd64
```

**Comment by glennmcallister** [comment]
With my modified `vault` binary:

```
$ vault auth -method=ldap username=glenn.mcallister
Password (will be hidden):
Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/ldap/login/glenn.mcallister
Code: 400. Errors:

* user is not a member of any authorized group; additionally, no LDAP groups found in groupDN 'OU=Cloud Ops,DC=bluecatlabs,DC=net'; only policies from locally-defined groups available
```

Note that the failure of the query is **completely** orthogonal to the issue. The issue was purely around being able to establish a TLS connection. I can connect now.

And my updated configuration:

```
$ vault read auth/ldap/config
Key            	Value
---            	-----
binddn         	<snip>
bindpass       	<snip>
certificate    	<snip>
discoverdn     	false
groupattr      	memberof
groupdn        	OU=Cloud Ops,DC=bluecatlabs,DC=net
groupfilter    	(&(objectClass=group)(member:1.2.840.113556.1.4.1941:={{.UserDN}}))
insecure_tls   	true
starttls       	false
tls_max_version	tls11
tls_min_version	tls11
upndomain      	bluecatlabs.net
url            	ldaps://<snip>.bluecatlabs.net
userattr       	samaccountname
userdn         	DC=bluecatlabs,DC=net

The following warnings were returned from the Vault server:
* Read access to this endpoint should be controlled via ACLs as it will return the configuration information as-is, including any passwords.
```

**Comment by glennmcallister** [comment]
With #2060 merged into master and targeted for 0.6.3, it's a reasonable workaround. While I agree re-deploying the certificate chain is the *purer* solution, it's a lot of certificates to deploy, and we'd have to basically rebuild the AD infrastructure from scratch, which is risky. MS doesn't make it easy, unfortunately. :( 

Thanks for the time and effort you've taking on this issue, @jefferai.

**Comment by jefferai** [comment]
Sure. Thanks for the PR!
